### PR TITLE
[SYCL] Disable persistent device code caching by default.

### PR DESCRIPTION
**Pull request by kbobrovs**
Per conversation with @pvchupin, having persistent cache on by default is
dangerous - i.e. unexpected writes to home or temp directory. Users should
opt-in instead via new SYCL_CACHE_ENABLE_PERSISTENT var.
SYCL_CACHE_DISABLE_PERSISTENT is deprecated.

Complementary test change: https://github.com/intel/llvm-test-suite/pull/313

Signed-off-by: kbobrovs <Konstantin.S.Bobrovsky@intel.com>

### CODE REVIEW

File: sycl/doc/EnvironmentVariables.md [hunk 1]
```diff
@@ -38,8 +38,9 @@ subject to change. Do not rely on these variables in production code.
 | `SYCL_ENABLE_PCI` | Integer | When set to 1, enables obtaining the GPU PCI address when using the Level Zero backend. The default is 0. |
 | `SYCL_HOST_UNIFIED_MEMORY` | Integer | Enforce host unified memory support or lack of it for the execution graph builder. If set to 0, it is enforced as not supported by all devices. If set to 1, it is enforced as supported by all devices. |
 | `SYCL_CACHE_DIR` | Path | Path to persistent cache root directory. Default values are `%AppData%\libsycl_cache` for Windows and `$XDG_CACHE_HOME/libsycl_cache` on Linux, if `XDG_CACHE_HOME` is not set then `$HOME/.cache/libsycl_cache`. |
-| `SYCL_CACHE_TRACE` | Any(\*) | If the variable is set, messages are sent to std::cerr when non-blocking failures happen (e.g. unable to access cache item file). |
-| `SYCL_CACHE_DISABLE_PERSISTENT` | Any(\*) | Switches persistent cache switch off when the variable is set. |
+| `SYCL_CACHE_TRACE` | Any(\*) | If the variable is set, messages are sent to std::cerr when caching events or non-blocking failures happen (e.g. unable to access cache item file). |
+| `SYCL_CACHE_DISABLE_PERSISTENT (deprecated)` | Any(\*) | Has no effect. |
```
Comment (pvchupin) [inline sycl/doc/EnvironmentVariables.md:42]: Does it make sense to rework this var in similar to SYCL_HOST_UNIFIED_MEMORY, which is basically bool?
Call it SYCL_CACHE_PERSISTENT. Make it zero = OFF and non-zero = ON. Default can change independently based on what we decide now and later?
Looks like better control over behavior and more stable as a variable.

Comment (kbobrovs) [reply sycl/doc/EnvironmentVariables.md:42]: addressed in 2964f91

**Comment by kbobrovs** [comment]
Some glitch in testing:
```
/localdisk2/sycl_ci/buildbot/worker/sycl-ubu-x64-pr/llvm.src/sycl/source/detail/persistent_device_code_cache.cpp:365:54: error: 'static const char* cl::sycl::detail::SYCLConfig<Config>::getName() [with cl::sycl::detail::ConfigID Config = (cl::sycl::detail::ConfigID)14]' is private within this context
           SYCLConfig<SYCL_CACHE_PERSISTENT>::getName() + std::string{": "} +
                                                      ^
In file included from /localdisk2/sycl_ci/buildbot/worker/sycl-ubu-x64-pr/llvm.src/sycl/source/detail/persistent_device_code_cache.hpp:16:0,
                 from /localdisk2/sycl_ci/buildbot/worker/sycl-ubu-x64-pr/llvm.src/sycl/source/detail/persistent_device_code_cache.cpp:11:
/localdisk2/sycl_ci/buildbot/worker/sycl-ubu-x64-pr/llvm.src/sycl/source/detail/config.hpp:117:22: note: declared private here
   static const char *getName() { return BaseT::MConfigName; }
                      ^~~~~~~
```
line numbers don't correspond to the patch. `getName()` is declared at line 110. Will try rebasing to fix this.

**Comment by kbobrovs** [comment]
> Some glitch in testing:

the problem was due to newly-added code above getName() by some other patch:

```
private:
  static const char *getCachedValue(bool ResetCache = false) {...}
```

**Comment by kbobrovs** [comment]
forced-pushed for cherry-picking to internal code base

File: sycl/doc/EnvironmentVariables.md [hunk 2]
```diff
@@ -38,8 +38,9 @@ subject to change. Do not rely on these variables in production code.
...
+| `SYCL_CACHE_PERSISTENT` | Any(\*) | Controls persistent device compiled code cache - setting to '1' turns it on, to '0' - off. When on, SYCL runtime will try to cache and reuse JIT-compiled binaries. Default is off. |
```
Comment (pvchupin) [inline sycl/doc/EnvironmentVariables.md:43]:
```suggestion
| `SYCL_CACHE_PERSISTENT` | Integer | Controls persistent device compiled code cache. Turns it on if set to '1' and turns it off if set to '0'. When cache is enabled SYCL runtime will try to cache and reuse JIT-compiled binaries. Default is off. |
```

**Review by pvchupin — DISMISSED** [review]

**Comment by kbobrovs** [comment]
**Jenkins/Win** failed with expected Unit test failures (fixed by the intel/llvm-test-suite#313)
`[2021-06-06T19:42:19.263Z] Exp run 3 case(s) failed. Failed case(s): [SYCL :: KernelAndProgram/basic.cpp, SYCL :: KernelAndProgram/cache_env_vars.cpp, SYCL :: KernelAndProgram/spec_consts.cpp]`

**Jenkins/Lin** failed with the same three (expected) with one unrelated ESIMD test failure - will file JIRA and exclude until triaged.
`[2021-06-06T19:30:58.822Z] Exp run 4 case(s) failed. Failed case(s): [SYCL :: ESIMD/noinline_call_from_func.cpp, SYCL :: KernelAndProgram/basic.cpp, SYCL :: KernelAndProgram/cache_env_vars.cpp, SYCL :: KernelAndProgram/spec_consts.cpp]`

**sycl-win-x64-pr** failed due to timeout:
```
command timed out: 2400 seconds without output running [b'python3', b'llvm_ci/intel/worker/tools/build.py', b'-t', b'check-sycl', b'-p', b'sycl', b'-s', b'lit', b'-n', b'15663', b'-b', b'pull/3891/head', b'-r', b'3891', b'-P', b'intel/llvm', b'-m', b'sycl-win-x64-pr', b'-e', b'73ce9da41afb7ab39f2fe3f494641faac6874948', b'-U', b'http://ci.llvm.intel.com:8010/#builders/18/builds/15663'], attempting to kill
program finished with exit code 1
elapsedTime=2531.101994
```
rebuilding. If this passes second time, PR can be merged

**Comment by kbobrovs** [comment]
OK, second testing on Windows failed again with timeout. I wonder it cab be caused by a potential probability that prior to this patch we tested cached binaries, which took less time to run? Need to see what is the actual test case which caused the timeout.

File: sycl/test/Unit/lit.cfg.py [hunk 3]
```diff
@@ -45,6 +45,10 @@
     # Disable device library fallback for unit tests by default.
     config.environment['SYCL_DEVICELIB_NO_FALLBACK'] = "1"
 
+# Unconditionally enable persistent device code caching, as it is off
+# by default, and some unit tests depend on it.
+config.environment['SYCL_CACHE_PERSISTENT'] = "1"
```
Comment (bader) [inline sycl/test/Unit/lit.cfg.py:50]: Is it possible to enable this feature only for tests requiring it?
Enabling optional feature for all the tests might hide problems in default workflow.

Comment (romanovvlad) [reply sycl/test/Unit/lit.cfg.py]: Done.

**Review by AGindinson — COMMENTED** [review]
As a side thought, could/should support for customizing the cache directory be added?

**Comment by romanovvlad** [comment]
> 
> 
> As a side thought, could/should support for customizing the cache directory be added?

It seems there is already support - SYCL_CACHE_DIR

**Comment by romanovvlad** [comment]
The `noinline_call_from_func` is not related to the patch.

**Review by pvchupin — APPROVED** [review]

File: sycl/source/detail/persistent_device_code_cache.cpp [hunk 4]
```diff
@@ -153,7 +155,11 @@ std::vector<std::vector<char>> PersistentDeviceCodeCache::getItemFromDisc(
         isCacheItemSrcEqual(FileName + ".src", Device, Img, SpecConsts,
                             BuildOptionsString)) {
       try {
-        return readBinaryDataFromFile(FileName + ".bin");
+        std::string FullFileName = FileName + ".bin";
+        std::vector<std::vector<char>> res =
```
Comment (v-klochkov) [inline sycl/source/detail/persistent_device_code_cache.cpp:159]: Not blocking comment: 'res' -> 'Res'

**Review by v-klochkov — APPROVED** [review]
Looks good to me. One test fails on both OpenCL and L0

**Comment by bader** [comment]
> Looks good to me. One test fails on both OpenCL and L0

This seems to be a real regression. Can someone check, please?
If so, please, address non-blocking comment from @v-klochkov together with the fix for this regression.

**Comment by pvchupin** [comment]
Talked to @kbobrovs, he will disable the test for now. Regression looks real, but it seems we uncovered something new.

File: sycl/source/detail/config.def [hunk 5]
```diff
@@ -24,6 +24,7 @@ CONFIG(SYCL_HOST_UNIFIED_MEMORY, 1, __SYCL_HOST_UNIFIED_MEMORY)
 CONFIG(SYCL_CACHE_DIR, 164, __SYCL_CACHE_DIR)
 CONFIG(SYCL_CACHE_TRACE, 1, __SYCL_CACHE_TRACE)
 CONFIG(SYCL_CACHE_DISABLE_PERSISTENT, 1, __SYCL_CACHE_DISABLE_PERSISTENT)
+CONFIG(SYCL_CACHE_PERSISTENT, 1, __SYCL_CACHE_PERSISTENT)
```
Comment (stanleygambarin) [inline sycl/source/detail/config.def:27]: if the default behavior is 'off', shouldn't this value be initialized to '0' ?

Comment (kbobrovs) [reply sycl/source/detail/config.def:27]: this is not default value, it is size of var in-memory rep

**Pull request merged**